File: languages/scala/tak.scala
// tak: the Takeuchi function - the function-call / recursion-overhead axis of the suite.
// Naive recursive tak(x,y,z): three recursive calls per non-base node, NO memoization, NO
// iterative/tail rewrite. It touches no arrays and allocates nothing - the ONLY thing it
// stresses is the cost of a function call + return + a couple of integer compares/decrements.
// The size n maps to the classic shape tak(3n, 2n, n).
//
// Checksum = the TOTAL number of calls (counted at entry, before the base test; eager
// evaluation means all three inner calls always run). Secondary = the returned value. All
// integer; values stay tiny (no overflow).
object Tak {
  // Module-level call counter (like C's `static long calls`).
  var calls: Long = 0L

  def tak(x: Int, y: Int, z: Int): Int = {
    calls += 1
    if (y < x) tak(tak(x - 1, y, z), tak(y - 1, z, x), tak(z - 1, x, y))
    else z
  }

  def main(args: Array[String]): Unit = {
    val n = if (args.nonEmpty) args(0).toInt else 6
    val r = tak(3 * n, 2 * n, n)
    println(calls)
    println(s"tak($n) = $r")
  }
}

File: languages/java/tak.java
// tak: the Takeuchi function - the function-call / recursion-overhead axis of the suite.
// Naive recursive tak(x,y,z): three recursive calls per non-base node, NO memoization, NO
// iterative rewrite. It touches no arrays and allocates nothing - the ONLY thing it stresses
// is the cost of a function call + return + a couple of integer compares/decrements. That
// isolates a dimension nothing else measures (binary-trees recurses too, but is dominated by
// heap allocation). The size n maps to the classic shape tak(3n, 2n, n).
//
// Checksum = the TOTAL number of calls (every entry counted, before the base test); secondary =
// the returned value. All integer; values stay tiny (no overflow). The call count is the strict
// correctness invariant (eager evaluation means all three inner calls always run).

class Tak {
    // Class-level counter - incremented at every entry to tak, before the base-case test.
    static long calls = 0L;

    static int tak(int x, int y, int z) {
        calls++;
        if (y < x) return tak(tak(x - 1, y, z), tak(y - 1, z, x), tak(z - 1, x, y));
        return z;
    }

    public static void main(String[] args) {
        int n = args.length > 0 ? Integer.parseInt(args[0]) : 6;
        int r = tak(3 * n, 2 * n, n);
        System.out.println(calls);
        System.out.println("tak(" + n + ") = " + r);
    }
}
